cp_diffdock_api.py

```diff
@@ -0,0 +1,261 @@
+
+    @staticmethod
+    def process_results(preppedRequest: PreppedRequest) -> List[DiffDockProtocol.Response]:
+        """
+        DiffDock produces sdf files for each docked pose, in directories by complex name.
+        This function gathers the resulting sdfs for each complex to be returned by the DiffDock API.
+        """
+        allComplexResults = []
+
+        for complexInfo in preppedRequest.complex_infos:
+            # Find the output directory for this complex
+            complexName, proteinUnit, ligandUnit = complexInfo
+            proteinName = proteinUnit.name
+            ligandName = ligandUnit.name
+            resultDir = preppedRequest.out_dir / complexName
+            if not os.path.isdir(resultDir):
+                noPosesError = "No poses generated."
+                if not proteinUnit.ok():
+                    noPosesError += f" protein prep failure: {proteinUnit.error}."
+                if not ligandUnit.ok():
+                    noPosesError += f" ligand prep failure: {ligandUnit.error}."
+                allComplexResults.append(DiffDockProtocol.Result(proteinName, ligandName, error=noPosesError))
+                continue
+
+            # Process the sdfs in the output directory
+            poses = []
+            for dir_entry in os.scandir(resultDir):
+                filepath = dir_entry.path
+                filename = os.path.basename(filepath)
+
+                # Regex to get rank and confidence
+                rankRegex = re.compile(r"rank(\d+)(_confidence(-?[0-9.]+))?\.sdf")
+                match = rankRegex.match(filename)
+
+                if match:
+                    rankStr = match.group(1)
+                    confidenceStr = match.group(3)
+                    rankInt = int(rankStr)
+                    confidenceFloat = -1
+                    if confidenceStr:
+                        confidenceFloat = float(confidenceStr)
+                    with open(filepath, 'r') as f:
+                        fileContent = f.read()
+                    poses.append(DiffDockProtocol.Pose(filename, fileContent, rankInt, confidenceFloat))
+
+            poses.sort(key=lambda pose: pose.rank)
+            allComplexResults.append(DiffDockProtocol.Result(proteinName, ligandName, poses=poses))
+
+        return DiffDockProtocol.Response.makeResults(allComplexResults)
+
+    @staticmethod
+    def make_request_csv(entries: List[PreppedComplex]) -> str:
+        header = "complex_name,protein_path,ligand_description,protein_sequence\n"
+        csv = ""
+        for entry in entries:
+            complex_name, protein_unit, ligand_unit = entry
+            log.info(f"Working with {complex_name=} {type(protein_unit)} {type(ligand_unit)}")
+            protein_name, protein_label, protein_file = protein_unit
+            ligand_name, ligand_label, ligand_file = ligand_unit
+            if protein_unit.ok() and ligand_unit.ok():
+                csv += f"{complex_name},{protein_file},{ligand_file},\n"
```

Comment:
Comment that the names were constructed without commas, so there is no need for more careful csv handling.